wgengine/magicsock/magicsock.go

```diff
@@ -1505,128 +1469,85 @@ func (c *Conn) noteRecvActivityFromEndpoint(e conn.Endpoint) {
-				c.bufferedIPv4Packet = append(c.bufferedIPv4Packet[:0], b[:um.n]...)
-			}
-			c.pconn4.SetReadDeadline(time.Time{})
-		case <-c.donec:
-			return 0, nil, errors.New("Conn closed")
-		}
-		var regionID int
-		n, regionID = dm.n, dm.regionID
-		ncopy := dm.copyBuf(b)
-		if ncopy != n {
-			err = fmt.Errorf("received DERP packet of length %d that's too big for WireGuard ReceiveIPv4 buf size %d", n, ncopy)
-			c.logf("magicsock: %v", err)
+	for {
+		n, pAddr, err := c.pconn6.ReadFrom(b)
+		if err != nil {
 			return 0, nil, err
 		}
-
-		ipp = netaddr.IPPort{IP: derpMagicIPAddr, Port: uint16(regionID)}
-		if c.handleDiscoMessage(b[:n], ipp) {
-			goto Top
+		if ep, ok := c.receiveIP(b[:n], pAddr.(*net.UDPAddr)); ok {
+			return n, ep, nil
 		}
+	}
+}
 
-		var (
-			didNoteRecvActivity bool
-			discoEp             *discoEndpoint
-			asEp                *addrSet
-		)
-		c.mu.Lock()
-		if dk, ok := c.discoOfNode[tailcfg.NodeKey(dm.src)]; ok {
-			discoEp = c.endpointOfDisco[dk]
-			// If we know about the node (it's in discoOfNode) but don't know about the
-			// endpoint, that's because it's an idle peer that doesn't yet exist in the
-			// wireguard config. So run the receive hook, if defined, which should
-			// create the wireguard peer.
-			if discoEp == nil && c.noteRecvActivity != nil {
-				didNoteRecvActivity = true
-				c.mu.Unlock()          // release lock before calling noteRecvActivity
-				c.noteRecvActivity(dk) // (calls back into CreateEndpoint)
-				// Now require the lock. No invariants need to be rechecked; just
-				// 1-2 map lookups follow that are harmless if, say, the peer has
-				// been deleted during this time.  In that case we'll treate it as a
-				// legacy pre-disco UDP receive and hand it to wireguard which'll
-				// likely just drop it.
-				c.mu.Lock()
-
-				discoEp = c.endpointOfDisco[dk]
-				c.logf("magicsock: DERP packet received from idle peer %v; created=%v", dm.src.ShortString(), ep != nil)
-			}
-		}
-		if !c.disableLegacy {
-			asEp = c.addrsByKey[dm.src]
-		}
-		c.mu.Unlock()
+func (c *Conn) derpPacketArrived() bool {
+	rc := atomic.LoadInt64(&c.derpRecvCountAtomic)
+	if rc != c.derpRecvCountLast {
+		c.derpRecvCountLast = rc
+		return true
+	}
+	return false
+}
 
-		if discoEp != nil {
-			ep = discoEp
-		} else if asEp != nil {
-			ep = asEp
-		} else {
-			key := wgkey.Key(dm.src)
-			c.logf("magicsock: DERP packet from unknown key: %s", key.ShortString())
-			ep = c.findEndpoint(ipp, addr, b[:n])
-			if ep == nil {
-				goto Top
+// ReceiveIPv4 is called by wireguard-go to receive an IPv4 packet.
+// In Tailscale's case, that packet might also arrive via DERP. A DERP packet arrival
+// aborts the pconn4 read deadline to make it fail.
+func (c *Conn) ReceiveIPv4(b []byte) (n int, ep conn.Endpoint, err error) {
+	for {
+		n, pAddr, err := c.pconn4.ReadFrom(b)
+		if err != nil {
+			// If the pconn4 read failed, the likely reason is a DERP reader received
+			// a packet and interrupted us.
```

Comment:
Add a note that it's okay to process a derp packet on _any_ error, not just deadline exceeded, since other errors will persist until the next read?

(I had to persuade myself that this was safe when reading this)

Reply:
Done.